[verification] HIGH: v3-ci.yml main CI failing — smoke-trajectory-graph-edges.mjs OOM crash JavaScript heap out of memory (ADR-130 P3)

## Summary

`scripts/smoke-trajectory-graph-edges.mjs` (graph trajectory hooks smoke, ADR-130 P3) crashes with JavaScript heap out-of-memory during TEST 2 (post-task reinforced-by edge write). Node's heap grows to 4050+ MB before GC gives up and the process is killed with SIGABRT (exit 134). TEST 1 passes cleanly.

## Impact

- CI pipeline on `main` has been continuously failing since at least 2026-06-02 (all 5 most recent runs are `failure`)
- ADR-130 P3 trajectory graph hook cannot be validated in CI
- If the memory growth reflects a real runtime leak (not just test-harness specific), production agent sessions using trajectory graph edge recording could be affected
- Unblocks only when the memory leak in TEST 2 is identified and fixed

## Expected

`smoke-trajectory-graph-edges.mjs` completes all test cases successfully:

```
PASS  1d: graph_edges has trajectory-caused row (count=1)
TEST 2: post-task writes reinforced-by edge
PASS  2x: ...
```

## Actual

```
PASS  1d: graph_edges has trajectory-caused row (count=1)

TEST 2: post-task writes reinforced-by edge
[2768:0xb255000]    29877 ms: Scavenge (interleaved) 4050.5 (4128.9) -> 4048.3 (4130.1) MB
[2768:0xb255000]    30108 ms: Scavenge (interleaved) 4051.7 (4130.1) -> 4050.5 (4150.6) MB

FATAL ERROR: Reached heap limit Allocation failed - JavaScript heap out of memory
Aborted (core dumped) node scripts/smoke-trajectory-graph-edges.mjs
##[error]Process completed with exit code 134.
```

## Command output

Failing job: **graph trajectory hooks smoke (ADR-130 P3)** (job id 79526307991, run id 26953932236)
Failing step: **Run trajectory graph edges smoke**

```
2026-06-04T13:15:52.0767426Z   PASS  1c: tool returns in <200ms (took 1ms)
2026-06-04T13:15:52.5803327Z   PASS  1d: graph_edges has trajectory-caused row (count=1)
2026-06-04T13:15:52.5805306Z
2026-06-04T13:15:52.5806601Z TEST 2: post-task writes reinforced-by edge
2026-06-04T13:16:21.7899691Z [2768:0xb255000]    29877 ms: Scavenge (interleaved) 4050.5 MB
2026-06-04T13:16:21.7905207Z [2768:0xb255000]    30108 ms: Scavenge (interleaved) 4051.7 MB
2026-06-04T13:16:21.7907422Z FATAL ERROR: Reached heap limit Allocation failed - JavaScript heap out of memory
2026-06-04T13:16:40.3406469Z Aborted (core dumped) node scripts/smoke-trajectory-graph-edges.mjs
2026-06-04T13:16:40.3421438Z ##[error]Process completed with exit code 134.
```

Time between TEST 2 start and crash: ~29 seconds.

## Recommended fix

1. Profile `smoke-trajectory-graph-edges.mjs` TEST 2 — likely a mock/stub or graph structure that accumulates in memory without disposal between test cases.
2. Add explicit cleanup between TEST 1 and TEST 2 (close DB handles, clear in-memory graph caches).
3. Add `--max-old-space-size=512` to the CI smoke invocation to surface leaks faster with a cleaner error.
4. Consider whether the trajectory graph reinforced-by edge writer accumulates all prior edges in memory during batch writes.

## Relevant report section

Check 6: Recent CI status on main — SEVERITY:HIGH

CI run: https://github.com/ruvnet/ruflo/actions/runs/26953932236
Job: https://github.com/ruvnet/ruflo/actions/runs/26953932236/job/79526307991

## Commit SHA

`d065b15927c6ba7318623e8af123e7980e4c6681`

## Timestamp UTC

2026-06-07T07:05:32Z

--- a/.github/workflows/v3-ci.yml
+++ b/.github/workflows/v3-ci.yml
@@ -2322,13 +2322,14 @@ jobs:
           pnpm install --frozen-lockfile
           pnpm -r build
       - name: Run trajectory graph edges smoke
-        # #2312 — TEST 2 (post-task → reinforced-by edge) OOMs at 6 GB+ inside
-        # the SONA endTrajectory + distillLearning chain (EWCConsolidator init
-        # is the likely allocator). The graph-edge contract being asserted is
-        # orthogonal to that path. Gate the step non-blocking until the real
-        # allocation profile is fixed; TEST 1 (trajectory-step → trajectory-
-        # caused edge) still runs and gates the same contract.
-        continue-on-error: true
+        # #2312 root cause (fixed): memory-bridge's rescueAgentdbEmbedder
+        # patched agentdb.embedder.embed to delegate to generateEmbedding —
+        # which is bridge-first, so the patched embed re-entered itself via
+        # bridgeGenerateEmbedding in an unbounded async cycle (heap OOM at
+        # the V8 limit, SIGABRT 134; SONA/EWC were innocent). The rescue now
+        # delegates to generateLocalEmbedding (a bridge-free leaf), so this
+        # smoke gates again at default heap — verified green in a node:22
+        # Linux container at 512 MB with a cold model cache.
         run: node scripts/smoke-trajectory-graph-edges.mjs
 
   graph-plugin-adapter-smoke:

--- a/scripts/smoke-trajectory-graph-edges.mjs
+++ b/scripts/smoke-trajectory-graph-edges.mjs
@@ -9,7 +9,7 @@
  * Usage: node scripts/smoke-trajectory-graph-edges.mjs
  */
 
-import { fileURLToPath } from 'url';
+import { fileURLToPath, pathToFileURL } from 'url';
 import * as path from 'path';
 import * as fs from 'fs';
 import * as os from 'os';
@@ -17,6 +17,10 @@ import * as os from 'os';
 const __dirname = path.dirname(fileURLToPath(import.meta.url));
 const projectRoot = path.resolve(__dirname, '..');
 const distBase = path.join(projectRoot, 'v3/@claude-flow/cli/dist/src');
+// Windows: absolute paths must be file:// URLs for dynamic import() — a bare
+// C:\... path fails with ERR_UNSUPPORTED_ESM_URL_SCHEME (hit while debugging
+// #2312 locally). No-op on POSIX.
+const distImport = (rel) => import(pathToFileURL(path.join(distBase, rel)).href);
 
 // sql.js is available in root node_modules (installed by the CI setup step via
 // `npm install --legacy-peer-deps --ignore-scripts` at the repo root).
@@ -72,10 +76,10 @@ console.log('\n[ADR-130 smoke] Phase 3 — SONA trajectory-to-graph hooks\n');
 async function testTrajectoryStep() {
   console.log('TEST 1: trajectory-step writes trajectory-caused edge');
   try {
-    const { initializeMemoryDatabase } = await import(path.join(distBase, 'memory/memory-initializer.js'));
+    const { initializeMemoryDatabase } = await distImport('memory/memory-initializer.js');
     await initializeMemoryDatabase({ dbPath, force: true });
 
-    const mod = await import(path.join(distBase, 'mcp-tools/hooks-tools.js'));
+    const mod = await distImport('mcp-tools/hooks-tools.js');
     const traj = mod.hooksTrajectoryStep ?? mod.allHooksTools?.find(t => t.name === 'hooks_intelligence_trajectory-step');
     if (!traj) { fail('1a', 'hooks_intelligence_trajectory-step not found'); return; }
 
@@ -107,7 +111,7 @@ async function testTrajectoryStep() {
 async function testPostTask() {
   console.log('\nTEST 2: post-task writes reinforced-by edge');
   try {
-    const mod = await import(path.join(distBase, 'mcp-tools/hooks-tools.js'));
+    const mod = await distImport('mcp-tools/hooks-tools.js');
     const postTask = mod.hooksPostTask ?? mod.allHooksTools?.find(t => t.name === 'hooks_post-task');
     if (!postTask) { fail('2a', 'hooks_post-task not found'); return; }
 
@@ -139,7 +143,7 @@ async function testPostTask() {
 async function testPostTaskFailure() {
   console.log('\nTEST 3: post-task with success=false does not write reinforced-by edge');
   try {
-    const mod = await import(path.join(distBase, 'mcp-tools/hooks-tools.js'));
+    const mod = await distImport('mcp-tools/hooks-tools.js');
     const postTask = mod.hooksPostTask ?? mod.allHooksTools?.find(t => t.name === 'hooks_post-task');
 
     const beforeCount = await countEdgesByRelation('reinforced-by');
@@ -165,7 +169,7 @@ async function testPostTaskFailure() {
 async function testNonBlocking() {
   console.log('\nTEST 4: async writes are non-blocking (<200ms tool response)');
   try {
-    const mod = await import(path.join(distBase, 'mcp-tools/hooks-tools.js'));
+    const mod = await distImport('mcp-tools/hooks-tools.js');
     const traj = mod.hooksTrajectoryStep ?? mod.allHooksTools?.find(t => t.name === 'hooks_intelligence_trajectory-step');
 
     const times = [];

--- a/v3/@claude-flow/cli/src/memory/memory-bridge.ts
+++ b/v3/@claude-flow/cli/src/memory/memory-bridge.ts
@@ -568,25 +568,41 @@ async function rescueAgentdbEmbedder(agentdb: { embedder?: { pipeline?: unknown;
   // own embed() does the right thing and we should not interpose.
   if (emb.pipeline) return;
 
-  type EmbedFn = (text: string) => Promise<{ embedding: number[] | Float32Array; dimensions: number; model: string }>;
-  let generateEmbedding: EmbedFn | null = null;
+  // #2312: delegate to generateLOCALEmbedding, NOT generateEmbedding.
+  // generateEmbedding is bridge-first — routing the rescued embed through it
+  // closes a cycle (generateEmbedding → bridgeGenerateEmbedding →
+  // embedder.embed [patched] → generateEmbedding → …) that allocated
+  // unboundedly via microtasks until V8 hit the heap limit (~4 GB on the CI
+  // runner, SIGABRT 134). The local chain never consults the bridge, so the
+  // rescued embedder is a leaf. If the running memory-initializer build
+  // predates generateLocalEmbedding, decline the rescue entirely — a stale
+  // pairing must fail safe, not recurse.
+  type EmbedFn = (text: string) => Promise<{ embedding: number[] | Float32Array; dimensions: number; model: string; backend?: 'onnx' | 'mock' }>;
+  let localEmbed: EmbedFn | null = null;
   try {
-    const mod = (await import('./memory-initializer.js')) as unknown as { generateEmbedding?: EmbedFn };
-    generateEmbedding = mod.generateEmbedding ?? null;
+    const mod = (await import('./memory-initializer.js')) as unknown as { generateLocalEmbedding?: EmbedFn };
+    localEmbed = mod.generateLocalEmbedding ?? null;
   } catch {
     return; // can't import the rescuer — leave the mock fallback alone
   }
-  if (!generateEmbedding) return;
-  const embed: EmbedFn = generateEmbedding;
+  if (!localEmbed) return;
+  const embed: EmbedFn = localEmbed;
 
-  // Probe once to confirm the rescuer actually returns real (non-zero) vectors.
-  // If our own fallback chain also dead-ends in mock embeddings, leave
-  // agentdb as-is rather than swapping one mock for another.
+  // Probe once to confirm the rescuer actually returns REAL ONNX vectors
+  // (#2312: the old probe only checked non-zero, which the deterministic
+  // hash fallback also satisfies — so it "rescued" agentdb's mock with our
+  // own mock and reported it as real). Require backend === 'onnx'.
   try {
     const probe = await embed('rescue-probe');
     const arr = probe?.embedding ? Array.from(probe.embedding as ArrayLike<number>) : [];
     const hasSignal = arr.length > 0 && arr.some((v: number) => Math.abs(v) > 1e-9);
-    if (!hasSignal) return;
+    if (!hasSignal || probe.backend !== 'onnx') {
+      // Local chain is also degraded — leave agentdb's embedder alone, but
+      // tag it so bridgeGenerateEmbedding's AUDIT-#3 isMock check reports
+      // backend='mock' truthfully instead of labeling mock vectors 'onnx'.
+      try { (emb as { backend?: string }).backend = 'mock'; } catch { /* frozen */ }
+      return;
+    }
   } catch {
     return;
   }

--- a/v3/@claude-flow/cli/src/memory/memory-initializer.ts
+++ b/v3/@claude-flow/cli/src/memory/memory-initializer.ts
@@ -1867,6 +1867,30 @@ export async function generateEmbedding(text: string): Promise<{
     }
   }
 
+  return generateLocalEmbedding(text);
+}
+
+/**
+ * Generate an embedding using ONLY the local model chain (transformers.js /
+ * ruvector ONNX / hash fallback) — never the AgentDB bridge.
+ *
+ * #2312: this MUST stay bridge-free. `memory-bridge.ts` rescues a degraded
+ * agentdb embedder by delegating to this module; if that delegation went
+ * through `generateEmbedding` (bridge-first), the call would re-enter the
+ * patched `agentdb.embedder.embed` and recurse unboundedly:
+ *
+ *   generateEmbedding → bridgeGenerateEmbedding → embedder.embed (patched)
+ *     → generateEmbedding → … (heap OOM at ~4 GB on CI, no stack overflow
+ *     because the cycle is async/microtask-driven)
+ *
+ * Keeping the local chain as its own export breaks that cycle structurally.
+ */
+export async function generateLocalEmbedding(text: string): Promise<{
+  embedding: number[];
+  dimensions: number;
+  model: string;
+  backend: 'onnx' | 'mock';
+}> {
   // Ensure model is loaded
   if (!embeddingModelState?.loaded) {
     await loadEmbeddingModel();

